docs: add pytorch integration guide. (#2711)
prefix-dev/pixi: docs/features/pytorch.md

- Install PyTorch using `pytorch` Conda channel (Legacy)

With these options you can choose the best way to install PyTorch based on your requirements.

## Installing from Conda-forge
You can install PyTorch using the `conda-forge` channel.
These are the community maintained builds of PyTorch.
You can make direct use of the Nvidia provided packages to make sure the packages can work together.

!!! note
    Currently not well-supported for Windows, but there is lots of work being done to get this working.
    Follow the work on the [feedstock](https://github.com/conda-forge/pytorch-cpu-feedstock)
!!! note
    Pixi uses the `system-requirements.cuda` to tell it can use the `cuda` packages.
    Without it, pixi will install the `cpu` versions.

=== "`pixi.toml`"
    ```toml title="Bare minimum conda-forge pytorch with cuda installation"
    --8<-- "docs/source_files/pixi_tomls/pytorch-conda-forge.toml:minimal"
    ```
=== "`pyproject.toml`"
    ```toml title="Bare minimum conda-forge pytorch with cuda installation"
    --8<-- "docs/source_files/pyproject_tomls/pytorch-conda-forge.toml:minimal"
    ```

> To deliberately install a specific version of the `cuda` packages you can depend on the `cuda-version` package which will then be interpreted by the other packages at resolve time.

Reviewer: ```suggestion
To deliberately install a specific version of the `cuda` packages you can depend on the `cuda-version` package which will then be interpreted by the other packages during resolution.
```

> To deliberately install a specific version of the `cuda` packages you can depend on the `cuda-version` package which will then be interpreted by the other packages at resolve time.

Reviewer: Maybe add a closed note, with an explanation why you need `cuda-version` and the system-requirements?

Author: Added some more information, hopefully that helps, I added a `note` for the system requirements

Reviewer: Hmm still not sure if I get it. So if I put the `system-requirements` to a lower cuda, why do I still get a higher cuda pytorch variant for example? I would just expect that to work.

Reviewer: So maybe it will be more accurate to say, that some packages like `PyTorch` (instead of cudatoolkit) do not use the cuda system-requirement but instead put a constraint on the cuda-version? And that is is somewhat unspecified what will happen if you specify only `__cuda`?

Author: System requirements are not real version specifications. You're describing the system pixi can solve for. 
None of the packages in this case actually use the cuda version and often only depend on `__cuda = "*"` Even though they are using a higher version than you specify. 

But it really shows what a bad design this is. We need to fix this ASAP. 

I'll take a look if I can make it more clear.

Reviewer: I do see a requirement for cudatoolkit though?

> To deliberately install a specific version of the `cuda` packages you can depend on the `cuda-version` package which will then be interpreted by the other packages at resolve time.

Reviewer: ```suggestion
To deliberately install a specific version of the `cuda` packages you can depend on the `cuda-version` package which will then be interpreted by the other packages during resolution.
```

Author: changed

Without specifying the `cuda-version` package, the latest version of cuda will be installed.
=== "`pixi.toml`"
    ```toml title="Add cuda version to the conda-forge pytorch installation"
    --8<-- "docs/source_files/pixi_tomls/pytorch-conda-forge.toml:cuda-version"
    ```
=== "`pyproject.toml`"
    ```toml title="Add cuda version to the conda-forge pytorch installation"
    --8<-- "docs/source_files/pyproject_tomls/pytorch-conda-forge.toml:cuda-version"
    ```

With `conda-forge` you can also install the `cpu` version of PyTorch.

> A common use-case is having two environments, one for CUDA machines and one for none-CUDA machines.

Reviewer: ```suggestion
A common use-case is having two environments, one for CUDA machines and one for non-CUDA machines.
```

=== "`pixi.toml`"
    ```toml title="Adding a cpu environment"
    --8<-- "docs/source_files/pixi_tomls/pytorch-conda-forge.toml:use-envs"
    ```
=== "`pyproject.toml`"
    ```toml title="Split into environments and add a CPU environment"
    --8<-- "docs/source_files/pyproject_tomls/pytorch-conda-forge.toml:use-envs"
    ```

Now you should be able to extend that with your dependencies and tasks.

Links to notable packages:

- [pytorch](https://prefix.dev/channels/conda-forge/packages/pytorch)
- [pytorch-cpu](https://prefix.dev/channels/conda-forge/packages/pytorch-cpu)
- [pytorch-gpu](https://prefix.dev/channels/conda-forge/packages/pytorch-gpu)
- [torchvision](https://prefix.dev/channels/conda-forge/packages/torchvision)
- [torchaudio](https://prefix.dev/channels/conda-forge/packages/torchaudio)
- [cuda-version](https://prefix.dev/channels/conda-forge/packages/cuda-version)

> ## Installing from PyPi

Reviewer: There have been reports where mixing certain conda packages that depended on PyTorch with PyPI pytorch seemed to be troublesome, maybe advise the user on that here.

Author: Added some more information, if you remember where those reports are, please send them in maybe I can elaborate some more based on them. But I can't find them.

Reviewer: This is one: https://discord.com/channels/1082332781146800168/1312759176363249735

It gets hairy pretty quickly 😄 🚀

Author: Tried to improve the overall documentation based on some of the points in that chat.

=== "`pixi.toml`"
    ```toml title="Install PyTorch from pypi"
    --8<-- "docs/source_files/pixi_tomls/pytorch-pypi.toml:minimal"
    ```
=== "`pyproject.toml`"
    ```toml title="Install PyTorch from pypi"
    --8<-- "docs/source_files/pyproject_tomls/pytorch-pypi.toml:minimal"
    ```

## Installing from PyTorch channel
!!! warning
    This depends on the [non-free](https://www.anaconda.com/blog/is-conda-free) `main` channel from Anaconda and mixing it with `conda-forge` can lead to conflicts.

!!! note
    This is the [legacy](https://dev-discuss.pytorch.org/t/pytorch-deprecation-of-conda-nightly-builds/2590) way of installing pytorch, this will not be updated to later versions as pytorch has discontinued their channel.

=== "`pixi.toml`"
    ```toml title="Install PyTorch from the PyTorch channel"
    --8<-- "docs/source_files/pixi_tomls/pytorch-from-pytorch-channel.toml:minimal"
    ```
=== "`pyproject.toml`"
    ```toml title="Install PyTorch from the PyTorch channel"
    --8<-- "docs/source_files/pyproject_tomls/pytorch-from-pytorch-channel.toml:minimal"
    ```

## Troubleshooting

> - You can test the installation with the following command:

Reviewer: Looking at the rendered version I think using ### or #### instead of bullets would be nicer.

Author: I currently wrote it as a summary of the implementations. So that's why I put it on the main level of the document which is `##`. We could split it up but I mainly added this to allow search to find this file when users have these issues.

Reviewer: Not sure if I misunderstand but I mean the bullets under the `## Troubleshooting`. So you would get something more like:

```
## Troubleshooting

#### How can I test the installation?

```

And maybe make all of them questions? Like more of FAQ style, wdyt?

Author: Ohh that might look nicer indeed. Let me try that.

Author: Changed that! Looks much better!

```shell
pixi run python -c "import torch; print(torch.__version__); print(torch.cuda.is_available())"
```
- You can ask pixi which version of CUDA it finds on your computer with `pixi info`.
```
> pixi info
...
Virtual packages: __unix=0=0
                : __linux=6.5.9=0
                : __cuda=12.5=0
...
```
- Installing `torch` from PyPI and other packages from Conda channels doesn't work.
  - The lowest level package in the dependency tree that uses a PyPI package demands that all later packages are also PyPI packages.
- Reasons for broken installations

>   - Using both `conda-forge` and `pytorch` channels, this can lead to conflicts. Choose one or the other.

Reviewer: Maybe mention again that the `pytorch` channel is deprecated?

Reviewer 2: I don't think the `pytorch` channel is deprecated.  However, mixing it with `conda-forge` is unsupported as it is supposed to be installed alongside `defaults` and `nvidia`.  In general, guaranteeing compatibility between packages compiled with different build-time dependencies requires careful checking.  I think that the way it is written is OK as it emphasises mixing different compiled software stacks can lead to trouble.

Author: This is now written differently.

- Not installing the GPU version of the `pytorch` package:
  - Using [conda-forge](./#installing-from-conda-forge): Use the `system-requirements.cuda` to tell pixi to install the `cuda` packages. And set the `cuda-version` package to the version you want to install.
  - Using [PyPI](./#installing-from-pypi): Make sure you are using the correct [index](./#pytorch-index) for the version you want to install.
- Not being able to solve the environment:
```
├─▶ failed to resolve pypi dependencies
╰─▶ Because only the following versions of torch are available:
      torch<=2.5.1
      torch==2.5.1+cpu
  and torch==2.5.1 has no wheels with a matching Python ABI tag, we can conclude that torch>=2.5.1,<2.5.1+cpu cannot be used.
  And because torch==2.5.1+cpu has no wheels with a matching platform tag and you require torch>=2.5.1, we can conclude that your requirements are
  unsatisfiable.
```
This error occurs when the ABI tag of the Python version doesn't match the wheels available on PyPI.
Fix this by lowering the `requires-python` or `python` dependency.

> At the time of writing Python 3.13 is not supported by PyTorch wheels yet.

Reviewer: ```suggestion
At the time of writing Python 3.13 is not supported by many PyTorch-dependent wheels yet.
```

# Pytorch Integration Guide

## Overview
This guide explains how to integrate PyTorch with `pixi`, it supports multiple ways of installing PyTorch.

- Install PyTorch using `conda-forge` Conda channel (Recommended)
- Install PyTorch using `pypi`, using our `uv`'s integration. (Most versions available)

> - Install PyTorch using `pytorch` Conda channel (Legacy)

Reviewer 2: This is the pytorch supported way - why are we writing it's "legacy"?

Reviewer 2: Please scratch that - read it below.

With these options you can choose the best way to install PyTorch based on your requirements.

## Installing from Conda-forge
You can install PyTorch using the `conda-forge` channel.

> These are the community maintained builds of PyTorch.

Reviewer 2: I'd specify these are the "conda-forge community" maintained builds.  The "pytorch community" maintains another set.

> You can make direct use of the Nvidia provided packages to make sure the packages can work together.

Reviewer 2: I'm not sure this is supported.
